lib/Doctrine/ORM/Mapping/Factory/UnderscoreNamingStrategy.php

```diff
@@ -100,18 +121,37 @@ public function joinColumnName($propertyName, $className = null)
      */
     public function joinTableName($sourceEntity, $targetEntity, $propertyName = null)
     {
-        return $this->classToTableName($sourceEntity) . '_' . $this->classToTableName($targetEntity);
+        return $this->_classToTableName($sourceEntity) . '_' . $this->_classToTableName($targetEntity);
     }
 
     /**
      * {@inheritdoc}
      */
     public function joinKeyColumnName($entityName, $referencedColumnName = null)
     {
-        return $this->classToTableName($entityName) . '_' .
+        return $this->_classToTableName($entityName) . '_' .
                 ($referencedColumnName ?: $this->referenceColumnName());
     }
 
+    /**
+     * @param string $string
+     * @param bool $pluralize
+     *
+     * @return string
+     */
+    private function _classToTableName($className, $pluralize = false)
+    {
+        if (strpos($className, '\\') !== false) {
+            $className = substr($className, strrpos($className, '\\') + 1);
+        }
+
+        if ($pluralize) {
+            $className = Inflector::pluralize($className);
+        }
+
+        return $this->underscore($className);
+    }
```

Comment:
Can we avoid this duplication of code?

Reply:
You mean something like protected method in default strategy and `return $this->underscore(parent::_classToTableName($className, $pluralize));` in underscore strategy, and inherit underscore from default ?
Is it really needed ?

Reply:
I mean a new dependency you would inject in the constructors of the existing strategies, and initialize to a default implementation if not provided.

Reply:
Well `$plural = false` as constructor argument - is not what you mean or what ?

Reply:
Oh, seems I got your point - remove "UnderscoreNamingStrategy" and add constructor parameter for default one like "$underscore = false" ? But it breaks BC very... much ) I think

Reply:
But... for BC we can make `parent::__construct` with proper parameters :smile: 
Do I understand you correctly ?

Reply:
Almost… I mean this:
```
    public function __construct(/*parentParams*/, ?ClassToTableNameTransformer $classToTableNameTransformer = null)
    {
        parent::__construct($parentParams);

        if ($classToTableNameTransformer) {
            $this->classToTableNameTransformer = $classToTableNameTransformer;

            return;
        }
        $this->classToTableNameTransformer = new DefaultClassToTableNameTransformer();
    }
```

Reply:
One more class just for "class for one more class" ? Java bite you :smile: